runtime/random_numbers.cu

```diff
@@ -51,18 +51,48 @@ __device__ uint4 philox(
   return output;
 }
 
+// This is a uniform double in the range (0, 1]
+__device__ double raw_uniform_double(unsigned int x, unsigned int y) {
+  constexpr double scale = 1.0 / (double)(1ll << 53);
+  const unsigned long long z =
+      (unsigned long long)x ^ ((unsigned long long)y << (53 - 32));
+  return (double)z * scale + 0.5 * scale;
+}
+
+// This is a uniform float in the range (0, 1]
+__device__ float raw_uniform_float(unsigned int x) {
+  constexpr float scale = (float)(1.0 / (double)(1ll << 32));
+  return (float)x * scale + 0.5f * scale;
+}
+
+// This is not actually __heq, but rather a bitwise identical check, i.e.
+// special cases for nans, subnormals, and denormalized floats are ignored.
+template <typename T>
+__device__ __inline__ bool identical(const T& a, const T& b) {
+  static_assert(sizeof(T) == 2);
+  return *(reinterpret_cast<const uint16_t*>(&a)) ==
+      *(reinterpret_cast<const uint16_t*>(&b));
+}
```

Comment:
It's probably not faster. I used `identical` for checking equality with 1.0 for which is sufficient. But its insufficient for the range check so I think I do need to write an __heq after all.

Reply:
Btw it would be great if we could just include cuda_fp16.h ...

Reply:
Oh, isn't `__heq` just work? `sin` just worked, so I wondered why `__heq` doesn't.

Reply:
Interesting that you got `sin` to work. I can't seem to get `sin` or `hsin` to work with a `__half` argument. Similarly `==` and `__heq` are not found. In `fp16_support.cu` we implement a bunch of stuff manually using PTX which is the approach I was about to use for `__heq` (copying from `cuda_fp16.hpp`), but this is not very maintainable. Instead, what if we kept a list of CTK header files that we could compile in. I believe we can include those as strings and use a filename to string mapping in the `nvrtcCreateProgram` call. This would let us include the _actual_ `cuda_*.h` files we need.

Reply:
> Interesting that you got `sin` to work. I can't seem to get `sin` or `hsin` to work with a `__half` argument. Similarly `==` and `__heq` are not found. In `fp16_support.cu` we implement a bunch of stuff manually using PTX which is the approach I was about to use for `__heq` (copying from `cuda_fp16.hpp`), but this is not very maintainable. Instead, what if we kept a list of CTK header files that we could compile in. I believe we can include those as strings and use a filename to string mapping in the `nvrtcCreateProgram` call. This would let us include the _actual_ `cuda_*.h` files we need.

I was mentioning to `sin` for float. It is also defined in a header, but we can just use it without including any header. So I am surprised that `__heq` does not just work.

Why do you think inlining PTX is not maintainable? Isn't it just a single line of this instruction?
https://docs.nvidia.com/cuda/parallel-thread-execution/index.html#comparison-and-selection-instructions-setp

Maintaining the entire header could include out compilation time, because there are lots of unused code. Also, one header could depend on other headers, so we might eventually added lots of unused codes.

Reply:
> Isn't it just a single line of this instruction?

At least in the case of bfloat, the comparison operators in PTX are done differently for sm90 vs sm80. This could change in the future and we will need to update it.

> Maintaining the entire header could include out compilation time, because there are lots of unused code. Also, one header could depend on other headers, so we might eventually added lots of unused codes.

Yeah, that's probably a good enough reason.

Reply:
I just pushed a change using the "real" `__heq`: https://github.com/NVIDIA/Fuser/pull/2355/files#diff-961121dcab5318a234d035a609e6a50d1642ce6e15b27ae6caf782ec56b5f4a2R285-R304

Reply:
> At least in the case of bfloat, the comparison operators in PTX are done differently for sm90 vs sm80. This could change in the future and we will need to update it.

🤬🤬🤬